src/PostgresOutbox/Subscriptions/Replication/EventDataMapper.cs

```diff
@@ -0,0 +1,68 @@
+using Npgsql;
+using Npgsql.Replication.PgOutput.Messages;
+using PostgresOutbox.Serialization;
+using PostgresOutbox.Subscriptions.ReplicationMessageHandlers;
+using System.Text.Json;
+using Npgsql.Replication.PgOutput;
+
+namespace PostgresOutbox.Subscriptions.Replication;
+
+internal sealed class EventDataMapper(ITypeResolver resolver): IReplicationDataMapper
+{
+    public async Task<IEnvelope> ReadFromReplication(InsertMessage insertMessage, CancellationToken ct)
+    {
+        var id = string.Empty;
+        var columnNumber = 0;
+        var eventTypeName = string.Empty;
+        await foreach (var column in insertMessage.NewRow)
+        {
+            try
+            {
+                switch (columnNumber)
+                {
+                    case 0:
+                        id = column.Kind == TupleDataKind.BinaryValue
+                            ? (await column.Get<long>(ct)).ToString()
+                            : await column.Get<string>(ct);
+                        break;
+                    case 1:
+                        using (var textReader = column.GetTextReader())
+                        {
+                            eventTypeName = await textReader.ReadToEndAsync(ct);
+                            break;
+                        }
+                    case 2 when column.GetDataTypeName().Equals("jsonb", StringComparison.OrdinalIgnoreCase):
+                    {
+                        var eventType = resolver.Resolve(eventTypeName);
+                        ArgumentNullException.ThrowIfNull(eventType, eventTypeName);
+                        return new OkEnvelope(await JsonSerialization.FromJsonAsync(eventType, column.GetStream(), resolver.SerializationContext, ct));
+                    }
+                }
+            }
+            catch (Exception ex) when (ex is ArgumentException or NotSupportedException or InvalidOperationException or JsonException)
+            {
+                return new KoEnvelope(ex,id);
```

Comment:
**SIDENOTE:** In the long term, this should also be behind `Polly` retry policy and stored in some dead-letter table.

Reply:
At this level we are catching unsolvable exceptions: think of malformed data or unexpected column data type.
I would use `Polly` at a higher level to deal with network glitch, when data are eventually sent on the wire.